File: snippets/fsharp/System/Type/Equals/EqualsEx1.fs
module EqualsEx1

// <Snippet1>
open System
open System.Reflection

let isEqualTo (t: Type) (inst: obj) =
    match inst with
    | :? Type as t2 ->
        printfn $"{t.Name} = {t2.Name}: {t.Equals t2}\n"
    | _ ->
        printfn "Cannot cast the argument to a type.\n"

do 
    let t = typeof<int>
    typeof<int>.GetTypeInfo()
    |> isEqualTo t

    typeof<String>
    |> isEqualTo t

    let t = typeof<obj>
    typeof<obj>
    |> isEqualTo t

    let t = typeof<ResizeArray<_>>.GetGenericTypeDefinition()
    let obj4: obj = (ResizeArray<String>()).GetType()
    isEqualTo t obj4

    let t = typeof<Type>
    let obj5: obj = null
    isEqualTo t obj5
// The example displays the following output:
//       Int32 = Int32: True
//       
//       Int32 = String: False
//       
//       Object = Object: True
//       
//       List`1 = List`1: False
//       
//       Cannot cast the argument to a type.
// </Snippet1>

File: snippets/csharp/System/Type/Equals/EqualsEx1.cs
// <Snippet1>
using System;
using System.Collections.Generic;
using System.Reflection;

public class Example
{
    public static void Main()
    {
        Type t = typeof(int);
        object obj1 = typeof(int).GetTypeInfo();
        IsEqualTo(t, obj1);

        object obj2 = typeof(string);
        IsEqualTo(t, obj2);

        t = typeof(object);
        object obj3 = typeof(object);
        IsEqualTo(t, obj3);

        t = typeof(List<>);
        object obj4 = (new List<string>()).GetType();
        IsEqualTo(t, obj4);

        t = typeof(Type);
        object obj5 = null;
        IsEqualTo(t, obj5);
    }

    private static void IsEqualTo(Type t, object inst)
    {
        Type t2 = inst as Type;
        if (t2 != null)
            Console.WriteLine($"{t.Name} = {t2.Name}: {t.Equals(t2)}");
        else
            Console.WriteLine("Cannot cast the argument to a type.");

        Console.WriteLine();
    }
}
// The example displays the following output:
//       Int32 = Int32: True
//
//       Int32 = String: False
//
//       Object = Object: True
//
//       List`1 = List`1: False
//
//       Cannot cast the argument to a type.
// </Snippet1>
